scripts/build.py

```diff
@@ -380,34 +502,63 @@ def hash_should_be_inserted(filepath):
                    in FILEPATHS_NOT_TO_RENAME)
 
 
-def copy_files_source_to_target(source, target, file_hashes):
-    """Copies all files in source directory to target directory, renames
-    them to include hash in their name.
+def should_file_be_built(filepath):
+    """Determines if the file should be built.
+        - JS files: Returns False if filepath matches with pattern in
+        JS_FILENAMES_TO_IGNORE, else returns True.
+        - Python files: Returns False if filepath ends with _test.py, else
+        returns True
+        - Other files: Returns False if filepath matches with pattern in
+        GENERAL_FILENAMES_TO_IGNORE, else returns True.
+
+    Args:
+        filepath: str. Path relative to file we are currently building.
+
+    Returns:
+        bool. True if filepath should be built, else False.
+    """
+    if filepath.endswith('.js'):
+        return not any(filepath.endswith(p) for p in JS_FILENAMES_TO_IGNORE)
+    elif filepath.endswith('_test.py'):
+        return False
+    else:
+        return not any(
+            filepath.endswith(p) for p in GENERAL_FILENAMES_TO_IGNORE)
+
+
+def generate_copy_tasks_to_copy_from_source_to_target(
+        source, target, file_hashes):
+    """Generate copy task for each file in source directory, excluding files
+    with extensions in FILE_EXTENSIONS_TO_IGNORE. Insert hash from hash dict
+    into the destination filename.
 
     Args:
         source: str. Path relative to /oppia directory of directory
             containing files and directories to be copied.
         target: str. Path relative to /oppia directory of directory where
             to copy the files and directories.
-        file_hashes: dict(str, str). Dictionary of file hashes.
+        file_hashes: dict(str, str). Dictionary with filepaths as keys and
+            hashes of file content as values.
+
+    Returns:
+        copy_tasks: deque(Thread). A deque that contains all copy tasks queued
+            to be processed.
     """
     print 'Processing %s' % os.path.join(os.getcwd(), source)
     print 'Copying into %s' % os.path.join(os.getcwd(), target)
-
-    ensure_directory_exists(target)
-    shutil.rmtree(target)
-    for root, dirs, files in os.walk(os.path.join(os.getcwd(), source)):
-        for directory in dirs:
+    copy_tasks = collections.deque()
+    for root, dirnames, filenames in os.walk(os.path.join(os.getcwd(), source)):
+        for directory in dirnames:
             print 'Copying %s' % os.path.join(root, directory)
 
-        for filename in files:
+        for filename in filenames:
             source_path = os.path.join(root, filename)
             if target in source_path:
                 continue
             if source not in source_path:
                 continue
 
-            # Ignore files with certain extensions.
+            # Python files should not be copied to final build directory.
```

Comment:
So this comment means that once we finish copying, no Python files should end up in `/build`. This is because we serve the whole `build` directory. User will be able to see files like
https://www.oppia.org/extensions/interactions/TextInput/protractor.js
(but imagine Python files in place of the protractor.js file) 
This comment was added later on by me and the filtering logic has been around long before this PR. Hopefully I am understanding things correctly.